src/values.ts

```diff
@@ -1,21 +1,166 @@
 /**
- * An Instant represents a fixed point in time (called "exact time"), without
- * regard to calendar or location, e.g. July 20, 1969, at 20:17 UTC.
+ * An wrapper around the Fauna `Time` type. It, represents a fixed point in time
+ * without regard to calendar or location, e.g. July 20, 1969, at 20:17 UTC.
  *
- * @remarks This class follows the TC39 proposal for `Temporal.Instant`. See the
- * proposal here: https://tc39.es/proposal-temporal/docs/index.html.
- * Since fauna has both `Time` and `Date` data types, we want to make a clear
- * distinction between them and avoid using the native `Date` class.
+ * @remarks The Javascript `Date` type most closely resembles a Fauna `Time`,
+ * not a Fauna `Date`. However, Fauna stores `Time` values with nanosecond
+ * precision, while Javascript `Date` values only have millisecond precision.
+ * This FaunaTime class preserves precision by storing the original string value
+ * and should be used whenever possible to pass `Time` values back to Fauna.
+ * Converting to a Javascript date before sending to Fauna could result in loss
+ * of precision.
+ *
+ * TODO: link to beta docs for `Time` type
  */
-class FaunaTime {}
+export class FaunaTime {
+  readonly #value: string;
+
+  /**
+   * @remarks constructor is private to enforce using factory functions
+   */
+  private constructor(isoString: string) {
+    this.#value = isoString;
+  }
+
+  /**
+   * Creates a new {@link FaunaTime} from an ISO date string
+   * @param item - An ISO date string.
+   * @returns A new {@link FaunaTime}
+   * @throws TypeError if a string is not provided, or RangeError if item
+   * is not a valid date
+   */
+  static from(item: string): FaunaTime {
+    if (typeof item !== "string") {
+      throw new TypeError(
+        `Expected string but received ${typeof item}: ${item}`
+      );
+    }
+    const validatedDate = new Date(item);
+    if (validatedDate.toString() === "Invalid Date") {
+      throw new RangeError(
+        `Expected a valid date string but received '${item}'`
+      );
+    }
+
+    return new FaunaTime(item);
+  }
+
+  /**
+   * Creates a new {@link FaunaTime} from a Javascript `Date`
+   * @param isoString - A Javascript `Date`
+   * @returns A new {@link FaunaTime}
+   */
+  static fromDate(date: Date): FaunaTime {
+    return new FaunaTime(date.toISOString());
+  }
+
+  /**
+   * Get the underlying string value
+   */
+  get value(): string {
+    return this.#value;
+  }
+
+  /**
+   * Get a copy of the `FaunaTime` converted to a Javascript `Date`. Does not
+   * mutate the existing `FaunaTime` value.
+   * @returns A `Date`
+   */
+  toDate(): Date {
+    return new Date(this.#value);
```

Comment:
the contract on the Date constructor is weak - it only supports a subset of ISO-8601.

Looks like parse: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/parse has a better contract.

So we could do:

```
return new Date(Date.parse(this.#value));
```

Reply:
Something like `"01 Jan 1970 00:00:00 GMT"` passes `Date.parse` but Fauna cannot accept it.

Reply:
yeah that's why i want to move off of it.

Reply:
we should only accept ISO-8601 strings for clarity.

Reply:
sorry i misread this. Let me read the MDN docs again.

Reply:
Ah I see the docs say parse only _explicitly_ supports ISO-8601 and that other format behavior is browser specific.

Yikes.

Looks like Date constructor has similar issues - it will take non ISO-8601.

I'd propose the following:

1. Let's maintain an ISO-8601 regex.
2. Verify the input matches the regex
3. Use parse to avoid Date constructor's pitfalls

Reply:
That's a plan!

Reply:
intersect that java date-parsing problems 🙄 Fauna _requires_ the timezone part

![image](https://user-images.githubusercontent.com/7998974/223848921-f287b129-4471-4d2f-8b87-5e0d3af387f4.png)

Reply:
clarification, spec assumes that no TZ means "local" time and so up to you to handle. I figured we'd let Fauna assume UTC. We'll have more explicitly convert to UTC or bounce back to the user. An error would be the simplest thing, with a message that says a Time must be relative to a timezone.

Reply:
sorry i don't quite follow your latest message here.

Reply:
Fauna requires all `Time` values to have a timezone part, e.g. `Z` or `+02:00` or `+2000`, etc.

If we _want_, when we construct a `TimeStub` directly from a string, we could catch that no timezone part was provided and assume UTC, i.e. append `"Z"` to the end.

The PR so far says if the string doesn't match then throw an error. It doesn't try anything more sophisticated. I think this is the right path. And we can always add support for parsing "local time" (i.e. ISO string with no TZ part) later.

Reply:
cool sounds right to me too